Adding 3-ary LongHeap to speed up collectors like TopDoc*Collectors

RamakrishnaChilaka opened the pull request:
### Description
This PR updates LongHeap from a fixed 2-ary heap to a 3-ary heap (the code is generic with n-ary Heap). The change improves cache locality and reduces heap operations for larger heaps, which improves overall query performance.

### Benchmarks
Luceneutil has been run with TopN as 1000 (see https://github.com/mikemccand/luceneutil/pull/357) Benchmark machine: i3.8xlarge

```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
     BrowseRandomLabelSSDVFacets        1.59     (13.0%)        1.52      (6.1%)   -4.4% ( -20% -   16%) 0.167
                HighSloppyPhrase       20.10      (9.3%)       19.52      (6.4%)   -2.9% ( -17% -   14%) 0.258
            BrowseDateSSDVFacets        0.51     (10.2%)        0.50     (10.4%)   -2.7% ( -21% -   20%) 0.415
     BrowseRandomLabelTaxoFacets        1.35     (15.1%)        1.32      (3.4%)   -2.4% ( -18% -   18%) 0.495
            BrowseDateTaxoFacets        1.83     (26.1%)        1.79      (9.0%)   -2.3% ( -29% -   44%) 0.705
               HighTermTitleSort       28.52      (2.9%)       27.98      (3.9%)   -1.9% (  -8% -    5%) 0.080
       BrowseDayOfYearTaxoFacets        1.88     (23.8%)        1.85      (9.3%)   -1.6% ( -27% -   41%) 0.777
            HighIntervalsOrdered        5.14      (6.0%)        5.08      (5.8%)   -1.2% ( -12% -   11%) 0.502
                     LowSpanNear       24.52      (6.3%)       24.29      (7.8%)   -0.9% ( -14% -   14%) 0.676
        AndHighHighDayTaxoFacets       19.61      (4.6%)       19.48      (5.2%)   -0.7% ( -10% -    9%) 0.673
           BrowseMonthTaxoFacets        1.96      (6.1%)        1.94      (6.3%)   -0.6% ( -12% -   12%) 0.748
            HighTermTitleBDVSort        6.95      (5.0%)        6.92      (5.8%)   -0.5% ( -10% -   10%) 0.785
                      TermDTSort       51.57      (4.6%)       51.36      (5.1%)   -0.4% (  -9% -    9%) 0.787
             LowIntervalsOrdered       72.91     (12.9%)       72.63     (15.0%)   -0.4% ( -25% -   31%) 0.933
             MedIntervalsOrdered        4.67      (5.2%)        4.66      (6.4%)   -0.3% ( -11% -   11%) 0.854
         AndHighMedDayTaxoFacets       27.57      (3.5%)       27.60      (3.9%)    0.1% (  -7% -    7%) 0.912
                 MedSloppyPhrase       11.49      (2.9%)       11.54      (2.7%)    0.4% (  -5% -    6%) 0.649
                      AndHighLow      676.59      (5.6%)      680.20      (4.4%)    0.5% (  -8% -   11%) 0.738
           HighTermDayOfYearSort       48.85      (6.8%)       49.18      (6.3%)    0.7% ( -11% -   14%) 0.748
                       LowPhrase       48.95      (4.5%)       49.52      (4.9%)    1.2% (  -7% -   11%) 0.429
          OrHighMedDayTaxoFacets        4.62      (7.1%)        4.68      (5.5%)    1.3% ( -10% -   14%) 0.512
                         Respell       21.00      (8.2%)       21.28      (6.3%)    1.3% ( -12% -   17%) 0.566
                 LowSloppyPhrase       18.02      (8.2%)       18.39      (6.0%)    2.0% ( -11% -   17%) 0.373
                      HighPhrase       26.79      (3.0%)       27.35      (2.5%)    2.1% (  -3% -    7%) 0.015
       BrowseDayOfYearSSDVFacets        2.44     (13.5%)        2.49     (10.8%)    2.2% ( -19% -   30%) 0.565
                          IntNRQ      124.80      (2.6%)      127.62      (2.7%)    2.3% (  -2% -    7%) 0.006
                     MedSpanNear       71.89      (3.3%)       73.63      (2.9%)    2.4% (  -3% -    8%) 0.014
                           range     2418.26      (7.8%)     2481.43      (6.2%)    2.6% ( -10% -   18%) 0.242
                          Fuzzy2       32.82      (6.2%)       33.72      (7.9%)    2.7% ( -10% -   17%) 0.222
                        PKLookup      124.86      (9.3%)      128.39      (9.8%)    2.8% ( -14% -   24%) 0.350
               HighTermMonthSort      109.23     (10.0%)      112.66     (11.5%)    3.1% ( -16% -   27%) 0.355
                    HighSpanNear       12.77     (10.1%)       13.23      (7.0%)    3.6% ( -12% -   23%) 0.189
            MedTermDayTaxoFacets       23.20      (7.2%)       24.08      (3.9%)    3.8% (  -6% -   16%) 0.039
                          IntSet      319.40      (7.9%)      332.99      (7.0%)    4.3% (  -9% -   20%) 0.070
                   OrHighNotHigh       63.45     (12.0%)       66.16     (11.1%)    4.3% ( -16% -   31%) 0.242
                    OrNotHighLow      653.14      (4.5%)      681.44      (5.5%)    4.3% (  -5% -   15%) 0.007
                          Fuzzy1       33.92      (9.6%)       35.47      (8.8%)    4.6% ( -12% -   25%) 0.116
                     AndHighHigh       72.73     (11.4%)       76.10     (12.0%)    4.6% ( -16% -   31%) 0.210
                        Wildcard      162.20      (3.9%)      169.92      (3.7%)    4.8% (  -2% -   12%) 0.000
                       MedPhrase      105.75      (5.3%)      111.28      (6.2%)    5.2% (  -5% -   17%) 0.004
                      AndHighMed      300.00      (3.5%)      317.31      (3.5%)    5.8% (  -1% -   13%) 0.000
                      OrHighHigh       81.99     (10.0%)       86.95     (11.8%)    6.1% ( -14% -   30%) 0.080
                   OrNotHighHigh       94.63      (8.0%)      100.89      (8.5%)    6.6% (  -9% -   25%) 0.011
                    OrNotHighMed      156.93      (5.2%)      167.46      (5.7%)    6.7% (  -3% -   18%) 0.000
                    OrHighNotLow      109.84     (12.4%)      118.15     (13.7%)    7.6% ( -16% -   38%) 0.067
                         Prefix3      298.72      (4.6%)      321.96      (4.8%)    7.8% (  -1% -   17%) 0.000
           BrowseMonthSSDVFacets        2.46      (8.4%)        2.66     (17.5%)    8.0% ( -16% -   37%) 0.066
                       OrHighMed      135.37      (5.3%)      147.30      (8.0%)    8.8% (  -4% -   23%) 0.000
                        HighTerm      169.55     (10.9%)      189.84     (15.3%)   12.0% ( -12% -   42%) 0.004
                       OrHighLow      182.04      (8.5%)      203.99     (10.4%)   12.1% (  -6% -   33%) 0.000
                         MedTerm      257.50      (9.8%)      297.02     (14.4%)   15.4% (  -8% -   43%) 0.000
                    OrHighNotMed      215.73      (6.5%)      249.31      (9.1%)   15.6% (   0% -   33%) 0.000
                         LowTerm      411.97      (8.6%)      481.02     (13.4%)   16.8% (  -4% -   42%) 0.000
```

jpountz commented:
Fascinating! Did you just confirm that 3 performs better than 2, or were you also able to confirm that 3 works better than 4 or 5?

jpountz commented:
Hmm I tried to reproduce with wikibigall on my machine, but luceneutil reports no speedup (nor slowdown).

RamakrishnaChilaka commented:
I ran additional benchmarks for arity=4 and arity=5 on an i3.8xlarge EC2 instance with `wikibigall` with topN as 1000. 
```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
           BrowseMonthTaxoFacets        2.11     (15.6%)        1.97      (8.6%)   -6.8% ( -26% -   20%) 0.090
                      AndHighLow      702.46      (4.8%)      679.44      (4.3%)   -3.3% ( -11% -    6%) 0.023
                           range     2588.42      (4.5%)     2510.42      (5.1%)   -3.0% ( -12% -    6%) 0.048
                        PKLookup      134.01      (9.4%)      130.17      (9.8%)   -2.9% ( -20% -   18%) 0.346
           BrowseMonthSSDVFacets        2.73     (14.2%)        2.68     (14.1%)   -1.9% ( -26% -   30%) 0.665
                          Fuzzy1       34.06      (4.1%)       33.45      (6.6%)   -1.8% ( -11% -    9%) 0.300
            HighIntervalsOrdered       11.29      (7.5%)       11.09      (6.9%)   -1.8% ( -15% -   13%) 0.438
            MedTermDayTaxoFacets       22.42      (3.6%)       22.03      (3.9%)   -1.8% (  -8% -    5%) 0.139
          OrHighMedDayTaxoFacets        7.03      (3.6%)        6.93      (4.0%)   -1.5% (  -8% -    6%) 0.205
                HighSloppyPhrase       11.25     (10.0%)       11.09     (12.5%)   -1.4% ( -21% -   23%) 0.703
       BrowseDayOfYearSSDVFacets        2.54      (7.4%)        2.50      (9.1%)   -1.3% ( -16% -   16%) 0.618
     BrowseRandomLabelSSDVFacets        1.62      (5.6%)        1.60      (6.4%)   -0.9% ( -12% -   11%) 0.642
     BrowseRandomLabelTaxoFacets        1.38      (3.6%)        1.37      (2.1%)   -0.8% (  -6% -    5%) 0.383
                       MedPhrase       27.46      (5.7%)       27.28      (6.3%)   -0.7% ( -11% -   12%) 0.727
             LowIntervalsOrdered      114.31     (10.3%)      113.73     (10.1%)   -0.5% ( -19% -   22%) 0.875
                      TermDTSort       49.04      (5.9%)       49.00      (4.6%)   -0.1% ( -10% -   11%) 0.969
               HighTermTitleSort       15.19      (2.3%)       15.21      (2.1%)    0.1% (  -4% -    4%) 0.853
            HighTermTitleBDVSort       19.89      (4.0%)       19.92      (2.1%)    0.2% (  -5% -    6%) 0.877
                         Respell       18.67      (6.9%)       18.70      (6.9%)    0.2% ( -12% -   15%) 0.928
         AndHighMedDayTaxoFacets       16.22      (2.9%)       16.26      (4.2%)    0.3% (  -6% -    7%) 0.804
                     MedSpanNear       14.25      (3.4%)       14.30      (3.6%)    0.4% (  -6% -    7%) 0.750
                    HighSpanNear       18.10      (5.6%)       18.17      (5.6%)    0.4% ( -10% -   12%) 0.822
        AndHighHighDayTaxoFacets       26.22      (3.2%)       26.38      (3.5%)    0.6% (  -5% -    7%) 0.560
            BrowseDateTaxoFacets        1.82      (9.3%)        1.84      (9.1%)    0.6% ( -16% -   21%) 0.830
                      HighPhrase       24.78      (5.0%)       24.99      (5.5%)    0.8% (  -9% -   11%) 0.621
                      OrHighHigh       54.07     (13.2%)       54.57     (14.9%)    0.9% ( -23% -   33%) 0.834
                 MedSloppyPhrase       40.49     (11.4%)       40.95      (8.0%)    1.1% ( -16% -   23%) 0.716
       BrowseDayOfYearTaxoFacets        1.87      (8.9%)        1.90      (8.6%)    1.6% ( -14% -   20%) 0.563
           HighTermDayOfYearSort       49.93      (4.7%)       50.76      (5.4%)    1.7% (  -8% -   12%) 0.298
                 LowSloppyPhrase       33.61      (4.2%)       34.34      (3.6%)    2.2% (  -5% -   10%) 0.076
                          IntSet      342.14      (5.0%)      350.17      (8.2%)    2.3% ( -10% -   16%) 0.274
                          Fuzzy2       38.24      (7.0%)       39.16      (8.8%)    2.4% ( -12% -   19%) 0.338
                       LowPhrase       81.34      (5.1%)       83.44      (6.7%)    2.6% (  -8% -   15%) 0.167
               HighTermMonthSort      124.62     (10.0%)      128.55      (8.6%)    3.2% ( -14% -   24%) 0.286
                   OrNotHighHigh      126.56      (9.3%)      131.11      (7.3%)    3.6% ( -11% -   22%) 0.173
                     AndHighHigh      110.96      (9.5%)      115.01     (10.4%)    3.7% ( -14% -   26%) 0.246
             MedIntervalsOrdered       36.48     (11.4%)       37.82     (10.3%)    3.7% ( -16% -   28%) 0.284
                      AndHighMed      331.48      (4.2%)      346.49      (3.3%)    4.5% (  -2% -   12%) 0.000
                   OrHighNotHigh       65.15     (10.0%)       68.40     (10.8%)    5.0% ( -14% -   28%) 0.130
                    OrHighNotMed      104.28     (11.8%)      109.63     (12.0%)    5.1% ( -16% -   32%) 0.173
                    OrNotHighMed      111.32      (8.0%)      117.42      (7.1%)    5.5% (  -8% -   22%) 0.021
                       OrHighLow      133.57     (11.5%)      141.84     (12.3%)    6.2% ( -15% -   33%) 0.099
                     LowSpanNear      116.39      (5.8%)      123.65      (9.6%)    6.2% (  -8% -   22%) 0.013
                          IntNRQ      401.26      (3.3%)      434.67      (4.3%)    8.3% (   0% -   16%) 0.000
            BrowseDateSSDVFacets        0.51     (11.3%)        0.55     (11.8%)    8.5% ( -13% -   35%) 0.021
                    OrHighNotLow      184.23     (12.8%)      200.08     (15.6%)    8.6% ( -17% -   42%) 0.057
                       OrHighMed      154.37      (6.9%)      167.71      (8.8%)    8.6% (  -6% -   26%) 0.001
                        HighTerm      193.83     (13.1%)      211.10     (18.2%)    8.9% ( -19% -   46%) 0.075
                        Wildcard      344.97      (6.0%)      376.99      (5.9%)    9.3% (  -2% -   22%) 0.000
                    OrNotHighLow      523.48      (5.6%)      572.78      (4.7%)    9.4% (   0% -   20%) 0.000
                         MedTerm      272.87     (12.8%)      303.38     (14.8%)   11.2% ( -14% -   44%) 0.011
                         Prefix3      346.15      (5.1%)      388.11      (7.0%)   12.1% (   0% -   25%) 0.000
                         LowTerm      424.14     (11.2%)      489.13     (17.5%)   15.3% ( -12% -   49%) 0.001
```

for n-ary 5
```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
          OrHighMedDayTaxoFacets        7.52      (6.1%)        7.22      (6.2%)   -3.9% ( -15% -    8%) 0.044
                      AndHighLow      918.22      (4.1%)      893.49      (3.3%)   -2.7% (  -9% -    4%) 0.022
                        PKLookup      131.07      (8.9%)      128.88     (13.7%)   -1.7% ( -22% -   22%) 0.647
     BrowseRandomLabelSSDVFacets        1.54      (8.9%)        1.51      (4.8%)   -1.6% ( -13% -   13%) 0.479
           HighTermDayOfYearSort       52.05      (5.2%)       51.47      (4.9%)   -1.1% ( -10% -    9%) 0.489
         AndHighMedDayTaxoFacets       71.07      (4.5%)       70.32      (6.2%)   -1.1% ( -11% -   10%) 0.535
                     LowSpanNear       23.05      (8.5%)       22.83      (7.8%)   -1.0% ( -15% -   16%) 0.703
           BrowseMonthTaxoFacets        1.96      (5.9%)        1.94      (8.1%)   -1.0% ( -14% -   13%) 0.664
        AndHighHighDayTaxoFacets       21.41      (4.8%)       21.21      (4.4%)   -1.0% (  -9% -    8%) 0.505
               HighTermTitleSort       21.28      (4.0%)       21.10      (3.7%)   -0.8% (  -8% -    7%) 0.507
                          Fuzzy1       45.08      (5.8%)       44.73      (8.4%)   -0.8% ( -14% -   14%) 0.731
            MedTermDayTaxoFacets       22.36      (5.2%)       22.23      (5.3%)   -0.6% ( -10% -   10%) 0.719
                 MedSloppyPhrase       18.25      (5.0%)       18.26      (6.0%)    0.0% ( -10% -   11%) 0.980
                HighSloppyPhrase       13.64      (6.5%)       13.72      (4.4%)    0.6% (  -9% -   12%) 0.737
                 LowSloppyPhrase       27.60      (6.7%)       27.78      (4.3%)    0.6% (  -9% -   12%) 0.713
       BrowseDayOfYearTaxoFacets        1.87      (8.5%)        1.88     (19.0%)    0.7% ( -24% -   30%) 0.886
                     MedSpanNear       14.28      (5.0%)       14.39      (4.4%)    0.7% (  -8% -   10%) 0.617
                      TermDTSort       51.11      (6.1%)       51.50      (7.6%)    0.8% ( -12% -   15%) 0.729
            HighIntervalsOrdered       10.27      (8.8%)       10.35      (9.4%)    0.8% ( -15% -   20%) 0.784
             LowIntervalsOrdered       43.86      (3.3%)       44.23      (3.4%)    0.9% (  -5% -    7%) 0.419
                       MedPhrase       67.55      (4.1%)       68.30      (4.5%)    1.1% (  -7% -   10%) 0.413
                    HighSpanNear       12.38      (2.4%)       12.53      (3.6%)    1.2% (  -4% -    7%) 0.214
     BrowseRandomLabelTaxoFacets        1.35      (3.6%)        1.37     (11.8%)    1.2% ( -13% -   17%) 0.652
            BrowseDateTaxoFacets        1.81      (9.8%)        1.84     (18.4%)    1.5% ( -24% -   32%) 0.742
               HighTermMonthSort      128.37     (11.1%)      130.37     (10.6%)    1.6% ( -18% -   26%) 0.650
             MedIntervalsOrdered       38.78     (10.2%)       39.43      (8.3%)    1.7% ( -15% -   22%) 0.564
                          Fuzzy2       35.57      (7.0%)       36.22      (8.7%)    1.8% ( -12% -   18%) 0.462
                           range     2331.03      (7.0%)     2380.31      (6.2%)    2.1% ( -10% -   16%) 0.312
                      HighPhrase       26.01      (2.2%)       26.60      (2.5%)    2.3% (  -2% -    7%) 0.002
                   OrNotHighHigh       69.94      (9.9%)       71.71     (10.4%)    2.5% ( -16% -   25%) 0.433
                          IntNRQ      251.37      (2.3%)      258.93      (2.7%)    3.0% (  -2% -    8%) 0.000
                   OrHighNotHigh       86.47      (8.6%)       89.21      (9.6%)    3.2% ( -13% -   23%) 0.271
                       LowPhrase      118.43      (3.0%)      122.61      (3.6%)    3.5% (  -3% -   10%) 0.001
            HighTermTitleBDVSort        7.52      (5.2%)        7.80      (5.8%)    3.7% (  -6% -   15%) 0.032
                     AndHighHigh       56.42      (9.7%)       58.54     (11.8%)    3.8% ( -16% -   27%) 0.272
           BrowseMonthSSDVFacets        2.59     (13.1%)        2.69     (23.1%)    3.8% ( -28% -   45%) 0.525
            BrowseDateSSDVFacets        0.52     (11.1%)        0.55     (10.5%)    4.4% ( -15% -   29%) 0.194
                         Respell       27.39      (8.8%)       28.64      (7.3%)    4.5% ( -10% -   22%) 0.075
                      AndHighMed      157.72      (6.9%)      165.18      (7.4%)    4.7% (  -8% -   20%) 0.036
                      OrHighHigh       59.00      (8.6%)       61.81      (7.1%)    4.8% ( -10% -   22%) 0.056
                    OrNotHighLow      509.65      (4.5%)      534.12      (6.1%)    4.8% (  -5% -   16%) 0.005
       BrowseDayOfYearSSDVFacets        2.51     (15.6%)        2.64     (20.1%)    5.2% ( -26% -   48%) 0.357
                          IntSet      350.80      (8.0%)      373.67      (6.4%)    6.5% (  -7% -   22%) 0.004
                        Wildcard      215.02      (4.5%)      230.84      (5.6%)    7.4% (  -2% -   18%) 0.000
                       OrHighMed      133.16      (5.9%)      144.00      (4.2%)    8.1% (  -1% -   19%) 0.000
                       OrHighLow      141.79      (8.9%)      153.40     (11.2%)    8.2% ( -10% -   31%) 0.010
                    OrNotHighMed      186.12      (4.9%)      201.54      (6.1%)    8.3% (  -2% -   20%) 0.000
                    OrHighNotMed      139.20      (8.5%)      150.98     (11.5%)    8.5% ( -10% -   31%) 0.008
                    OrHighNotLow      163.68      (8.6%)      181.52     (10.7%)   10.9% (  -7% -   33%) 0.000
                         Prefix3      409.57      (6.0%)      461.76      (7.9%)   12.7% (  -1% -   28%) 0.000
                        HighTerm      210.94      (9.6%)      242.76     (10.5%)   15.1% (  -4% -   38%) 0.000
                         MedTerm      258.25      (7.6%)      305.79      (8.9%)   18.4% (   1% -   37%) 0.000
                         LowTerm      420.58      (9.4%)      510.49     (10.7%)   21.4% (   1% -   45%) 0.000

```

RamakrishnaChilaka commented:
Reran benchmarks with arity 3
```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                        PKLookup      112.57      (9.9%)      105.75      (7.8%)   -6.1% ( -21% -   12%) 0.032
           BrowseMonthTaxoFacets        1.55      (9.8%)        1.51      (9.8%)   -2.9% ( -20% -   18%) 0.353
               HighTermTitleSort       21.94      (7.6%)       21.35      (9.7%)   -2.7% ( -18% -   15%) 0.328
                           range     1988.08      (7.4%)     1937.90      (7.5%)   -2.5% ( -16% -   13%) 0.284
                      AndHighLow      578.51      (6.5%)      566.10      (6.6%)   -2.1% ( -14% -   11%) 0.301
                HighSloppyPhrase        0.47      (6.7%)        0.46      (7.0%)   -1.7% ( -14% -   12%) 0.419
          OrHighMedDayTaxoFacets        2.07      (7.6%)        2.04      (6.0%)   -1.5% ( -14% -   13%) 0.486
     BrowseRandomLabelTaxoFacets        1.11      (5.9%)        1.10      (7.6%)   -0.6% ( -13% -   13%) 0.771
                         Respell       18.55      (7.2%)       18.44      (4.9%)   -0.6% ( -11% -   12%) 0.772
            MedTermDayTaxoFacets       19.47      (3.2%)       19.37      (3.0%)   -0.5% (  -6% -    5%) 0.580
                      HighPhrase       13.08      (4.8%)       13.06      (5.1%)   -0.1% (  -9% -   10%) 0.924
                     LowSpanNear       33.87      (4.7%)       33.85      (4.9%)   -0.1% (  -9% -    9%) 0.963
         AndHighMedDayTaxoFacets      107.22      (3.9%)      107.31      (2.9%)    0.1% (  -6% -    7%) 0.939
            HighTermTitleBDVSort       14.65      (4.0%)       14.66      (4.3%)    0.1% (  -7% -    8%) 0.931
           HighTermDayOfYearSort       45.00      (5.1%)       45.06      (5.5%)    0.1% ( -10% -   11%) 0.935
            HighIntervalsOrdered       18.87      (7.1%)       18.90      (8.5%)    0.1% ( -14% -   16%) 0.953
        AndHighHighDayTaxoFacets        6.24      (5.1%)        6.25      (4.5%)    0.2% (  -8% -   10%) 0.901
                      TermDTSort       43.90      (6.8%)       44.17      (7.5%)    0.6% ( -12% -   16%) 0.787
       BrowseDayOfYearTaxoFacets        1.50      (8.7%)        1.52     (12.0%)    1.3% ( -17% -   24%) 0.699
                 LowSloppyPhrase       49.24      (3.8%)       49.90      (6.3%)    1.3% (  -8% -   11%) 0.415
             LowIntervalsOrdered      108.45      (5.2%)      109.91      (7.4%)    1.3% ( -10% -   14%) 0.507
                     AndHighHigh       80.95     (11.3%)       82.26     (12.0%)    1.6% ( -19% -   28%) 0.662
                      OrHighHigh       64.43     (12.6%)       65.50     (14.3%)    1.7% ( -22% -   32%) 0.698
                    HighSpanNear       10.63      (4.7%)       10.83      (5.3%)    1.8% (  -7% -   12%) 0.254
                 MedSloppyPhrase       39.59     (14.4%)       40.35     (11.9%)    1.9% ( -21% -   33%) 0.649
                          IntNRQ      198.78      (4.5%)      202.61      (7.0%)    1.9% (  -9% -   14%) 0.302
                       MedPhrase       56.91      (2.7%)       58.01      (4.9%)    1.9% (  -5% -    9%) 0.121
            BrowseDateSSDVFacets        0.41     (11.6%)        0.42     (10.8%)    2.0% ( -18% -   27%) 0.572
             MedIntervalsOrdered       36.71      (4.5%)       37.45      (4.4%)    2.0% (  -6% -   11%) 0.154
     BrowseRandomLabelSSDVFacets        1.27      (7.6%)        1.31      (8.7%)    2.6% ( -12% -   20%) 0.308
                        Wildcard      161.19      (7.4%)      165.48      (8.2%)    2.7% ( -12% -   19%) 0.280
                         Prefix3      219.37      (5.6%)      225.28      (6.3%)    2.7% (  -8% -   15%) 0.155
       BrowseDayOfYearSSDVFacets        2.08     (14.6%)        2.14     (13.1%)    2.7% ( -21% -   35%) 0.533
                    OrNotHighMed       60.56      (7.0%)       62.25      (6.8%)    2.8% ( -10% -   17%) 0.202
                     MedSpanNear       81.72      (3.7%)       84.06      (4.8%)    2.9% (  -5% -   11%) 0.035
                          Fuzzy2       21.39      (4.1%)       22.02      (5.1%)    2.9% (  -5% -   12%) 0.044
                          Fuzzy1       27.20      (6.0%)       28.05     (10.3%)    3.1% ( -12% -   20%) 0.242
                   OrHighNotHigh       80.30     (11.7%)       83.06     (11.6%)    3.4% ( -17% -   30%) 0.353
               HighTermMonthSort       85.65      (7.5%)       88.62     (11.8%)    3.5% ( -14% -   24%) 0.268
                      AndHighMed      204.92     (10.9%)      212.49     (11.4%)    3.7% ( -16% -   29%) 0.295
                   OrNotHighHigh       94.39      (8.5%)       97.93      (9.9%)    3.7% ( -13% -   24%) 0.200
                       LowPhrase      121.30      (3.8%)      126.06      (5.2%)    3.9% (  -4% -   13%) 0.006
                    OrHighNotMed      134.97     (12.0%)      140.59     (13.2%)    4.2% ( -18% -   33%) 0.296
           BrowseMonthSSDVFacets        2.11     (11.5%)        2.23     (17.1%)    5.7% ( -20% -   38%) 0.219
            BrowseDateTaxoFacets        1.40      (5.9%)        1.49     (15.8%)    5.8% ( -14% -   29%) 0.123
                       OrHighLow      122.42      (9.9%)      129.57     (12.4%)    5.8% ( -14% -   31%) 0.099
                          IntSet      276.23      (7.6%)      293.37      (7.9%)    6.2% (  -8% -   23%) 0.011
                         MedTerm      174.33     (14.5%)      187.06     (17.4%)    7.3% ( -21% -   45%) 0.149
                    OrNotHighLow      355.78      (6.5%)      382.90      (6.5%)    7.6% (  -5% -   22%) 0.000
                        HighTerm      161.21     (14.0%)      173.70     (17.6%)    7.8% ( -20% -   45%) 0.124
                       OrHighMed      129.57      (7.2%)      139.74      (9.6%)    7.8% (  -8% -   26%) 0.003
                    OrHighNotLow      208.81      (7.5%)      236.41     (12.8%)   13.2% (  -6% -   36%) 0.000
                         LowTerm      352.74     (13.3%)      399.51     (17.0%)   13.3% ( -15% -   50%) 0.006
```

jpountz reviewed (commented):
Thank you, these benchmark results are quite convincing. I'm slightly uneasy about introducing loops that are unlikely to be unrolled, I could see it cause performance degradations in cases that we are not benchmarking here. Maybe keep `LongHeap` binary for now and introduce a `TernaryLongHeap` class? We can share code via `static void upHeap(long[] heap, int i, int arity)` helpers, and passing constants as the arity should help the compiler unroll the loops when it gets inlined? Let's only move top-hits collection to `TernaryLongHeap` for now and look into whether it also makes sense for vector search in a follow-up? What do you think?

RamakrishnaChilaka commented:
Thanks @jpountz for reviewing this PR.

I agree that it makes sense to keep `LongHeap` binary for now, given the scope of our current benchmarks. As you suggested, I’ll refactor the shared logic into static helpers `upHeap`/`downHeap` that take `arity` as a constant, introduce a `TernaryLongHeap` class with `arity=3` hardcoded, and switch only top-hits collectors to use it in this PR.

We can then revisit vector search in a follow-up once we have benchmarks for that workload to validate whether it benefits from the same change.

RamakrishnaChilaka commented:
I have implemented the following changes:

* Refactored `LongHeap` to include static helper methods `upHeap()` and `downHeap()` that accept arity as a parameter
* Added a new `TernaryLongHeap` class with `arity = 3`.
* Updated `TopScoreDocCollector` to use `TernaryLongHeap` instead of `LongHeap`.
* Left `LongHeap` as binary-arity (`arity = 2`) unchanged, i.e. its `upHeap`/`downHeap` remain separate to avoid any regressions. 


I re-ran benchmarks on c8g.8xlarge (graviton) instance.

```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
           BrowseMonthTaxoFacets        3.21      (6.7%)        3.14      (2.0%)   -2.3% ( -10% -    6%) 0.133
            BrowseDateTaxoFacets        3.94     (11.0%)        3.88      (8.0%)   -1.6% ( -18% -   19%) 0.587
       BrowseDayOfYearTaxoFacets        3.95     (10.6%)        3.88      (7.7%)   -1.6% ( -17% -   18%) 0.581
                      AndHighLow     1644.06      (3.0%)     1618.80      (1.9%)   -1.5% (  -6% -    3%) 0.053
           HighTermDayOfYearSort      113.88      (3.4%)      112.21      (2.8%)   -1.5% (  -7% -    4%) 0.137
            BrowseDateSSDVFacets        1.06      (9.3%)        1.05      (9.7%)   -1.1% ( -18% -   19%) 0.719
     BrowseRandomLabelTaxoFacets        2.95      (7.9%)        2.92      (5.7%)   -1.0% ( -13% -   13%) 0.633
                      TermDTSort      140.32      (3.1%)      138.94      (3.6%)   -1.0% (  -7% -    5%) 0.357
            HighTermTitleBDVSort       34.29      (2.3%)       33.96      (2.5%)   -0.9% (  -5% -    3%) 0.212
                           range     4030.14      (4.9%)     4004.47      (3.4%)   -0.6% (  -8% -    8%) 0.633
            MedTermDayTaxoFacets       58.42      (1.6%)       58.08      (2.1%)   -0.6% (  -4% -    3%) 0.312
         AndHighMedDayTaxoFacets       62.61      (0.9%)       62.30      (1.0%)   -0.5% (  -2% -    1%) 0.101
               HighTermTitleSort       86.93      (2.0%)       86.73      (2.7%)   -0.2% (  -4% -    4%) 0.758
                HighSloppyPhrase       32.36      (3.1%)       32.34      (3.1%)   -0.1% (  -6% -    6%) 0.932
     BrowseRandomLabelSSDVFacets        4.08     (10.8%)        4.08     (11.0%)    0.0% ( -19% -   24%) 0.999
       BrowseDayOfYearSSDVFacets        5.26      (7.0%)        5.27     (12.0%)    0.2% ( -17% -   20%) 0.954
        AndHighHighDayTaxoFacets       26.54      (0.9%)       26.59      (1.2%)    0.2% (  -1% -    2%) 0.571
                        PKLookup      245.99      (1.2%)      246.64      (2.4%)    0.3% (  -3% -    3%) 0.657
                 MedSloppyPhrase       76.81      (3.0%)       77.12      (3.6%)    0.4% (  -6% -    7%) 0.695
          OrHighMedDayTaxoFacets       12.97      (4.0%)       13.02      (4.5%)    0.5% (  -7% -    9%) 0.739
                         Respell       38.33      (2.0%)       38.55      (1.9%)    0.6% (  -3% -    4%) 0.329
            HighIntervalsOrdered       10.47      (9.0%)       10.53      (9.7%)    0.6% ( -16% -   21%) 0.838
                    HighSpanNear        5.59      (3.6%)        5.63      (3.2%)    0.8% (  -5% -    7%) 0.468
               HighTermMonthSort      252.82      (3.9%)      254.86      (5.5%)    0.8% (  -8% -   10%) 0.592
                       LowPhrase       77.26      (1.8%)       78.11      (1.3%)    1.1% (  -2% -    4%) 0.028
                 LowSloppyPhrase      131.76      (2.5%)      133.56      (2.3%)    1.4% (  -3% -    6%) 0.075
             MedIntervalsOrdered       53.50      (4.6%)       54.25      (4.9%)    1.4% (  -7% -   11%) 0.349
                          Fuzzy1       93.04      (1.1%)       94.40      (1.4%)    1.5% (  -1% -    4%) 0.000
           BrowseMonthSSDVFacets        5.53     (13.0%)        5.62     (15.0%)    1.5% ( -23% -   33%) 0.736
                          Fuzzy2       79.01      (1.3%)       80.34      (1.4%)    1.7% (   0% -    4%) 0.000
                      HighPhrase       52.31      (1.7%)       53.19      (1.5%)    1.7% (  -1% -    4%) 0.001
                   OrNotHighHigh      178.27      (6.7%)      181.77      (6.4%)    2.0% ( -10% -   16%) 0.344
                       MedPhrase       93.34      (1.7%)       95.35      (1.0%)    2.2% (   0% -    4%) 0.000
                   OrHighNotHigh      161.33      (7.8%)      165.05      (7.7%)    2.3% ( -12% -   19%) 0.345
             LowIntervalsOrdered      113.89      (6.3%)      116.73      (7.0%)    2.5% ( -10% -   16%) 0.234
                         Prefix3      446.66      (1.1%)      460.53      (2.2%)    3.1% (   0% -    6%) 0.000
                          IntSet      797.35      (3.6%)      829.10      (5.9%)    4.0% (  -5% -   14%) 0.010
                     LowSpanNear      292.44      (3.4%)      304.43      (2.7%)    4.1% (  -1% -   10%) 0.000
                     AndHighHigh      171.85     (12.6%)      179.11     (12.5%)    4.2% ( -18% -   33%) 0.288
                          IntNRQ      540.65      (3.0%)      564.98      (1.8%)    4.5% (   0% -    9%) 0.000
                      OrHighHigh      145.55     (10.5%)      152.50     (12.0%)    4.8% ( -16% -   30%) 0.180
                        Wildcard      691.41      (2.7%)      728.49      (2.9%)    5.4% (   0% -   11%) 0.000
                    OrNotHighLow     1070.20      (1.7%)     1127.88      (2.8%)    5.4% (   0% -   10%) 0.000
                     MedSpanNear      223.45      (2.3%)      235.74      (1.8%)    5.5% (   1% -    9%) 0.000
                      AndHighMed      460.42      (4.5%)      486.06      (4.6%)    5.6% (  -3% -   15%) 0.000
                    OrNotHighMed      381.97      (3.9%)      404.81      (4.5%)    6.0% (  -2% -   14%) 0.000
                       OrHighLow      275.73     (10.0%)      298.67     (11.7%)    8.3% ( -12% -   33%) 0.015
                        HighTerm      270.46     (13.9%)      294.96     (16.8%)    9.1% ( -18% -   46%) 0.063
                    OrHighNotMed      387.26      (6.6%)      425.74      (8.6%)    9.9% (  -4% -   26%) 0.000
                    OrHighNotLow      367.92     (11.3%)      406.06     (15.2%)   10.4% ( -14% -   41%) 0.014
                       OrHighMed      384.93      (4.2%)      427.50      (5.5%)   11.1% (   1% -   21%) 0.000
                         MedTerm      533.01      (7.4%)      611.25      (9.9%)   14.7% (  -2% -   34%) 0.000
                         LowTerm      779.05      (7.2%)      917.16      (9.3%)   17.7% (   1% -   36%) 0.000
```

jpountz commented on lucene/CHANGES.txt:
```diff
@@ -130,6 +130,8 @@ API Changes
   instance instead of a Bits instance to identify document IDs to filter.
   (Shubham Chaudhary, Adrien Grand)
 
+* GITHUB#15140: Optimize TopScoreDocCollector with TernaryLongHeap for improved performance over Binary-LongHeap. (Ramakrishna Chilaka)
```
Can you move it to the 10.4 section now that the 10.4 branch has been cut?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/TernaryLongHeap.java:
```diff
@@ -0,0 +1,187 @@
+    heap[size] = element;
+    LongHeap.upHeap(heap, size, 3);
+    return heap[1];
+  }
```
I had not realized that LongHeap grows unbounded on-demand. This doesn't feel right when the ctor takes a `maxSize` parameter. Not related to your PR but we should fix it in a separate PR?

jpountz replied:
> Rename maxSize parameter to initialSize/initialCapacity to better reflect its behaviour.

+1

jpountz commented on lucene/core/src/java/org/apache/lucene/util/TernaryLongHeap.java:
```diff
@@ -0,0 +1,187 @@
+      long result = heap[1]; // save first value
+      heap[1] = heap[size]; // move last to first
+      size--;
+      LongHeap.downHeap(heap, 1, size, 3); // adjust heap
```
nit: let's have a `ARITY = 3` constant instead of inlining 3?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/LongHeap.java:
```diff
@@ -216,4 +216,63 @@ public long get(int i) {
+      i = best;
+    }
+    heap[i] = value; // install saved value
+  }
```
Let's move these helpers to `TernaryLongHeap` since they're not used here and only used there?

RamakrishnaChilaka replied:
I initially put the helpers in LongHeap since the name is more generic, but I agree that keeping them in `TernaryLongHeap` makes sense for encapsulation and avoids unused code (for LongHeap) in LongHeap's class. I’ll move them over.

jpountz reviewed (commented):
I left minor comments but the change looks good to me overall!

RamakrishnaChilaka committed:
Optimize TopScoreDocCollector with TernaryLongHeap

RamakrishnaChilaka committed:
rename maxSize to initialCapacity in LongHeap and TernaryLongHeap

RamakrishnaChilaka commented:
<img width="1719" height="725" alt="image" src="https://github.com/user-attachments/assets/231754e4-a9e7-4a73-9914-80c03aeb773c" />


https://benchmarks.mikemccandless.com/Term.html
https://benchmarks.mikemccandless.com/2025.09.02.18.04.31.html

Latest runs shows improved performance for Term query

RamakrishnaChilaka commented:
I ran benchmarks with arity=3 and topN=20 on wikibigall on i3.8xlarge

The results show no statistically significant regressions with topN as 20.


```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                    OrHighNotMed      560.65     (12.3%)      532.49     (10.4%)   -5.0% ( -24% -   20%) 0.164
           HighTermDayOfYearSort      389.71      (6.1%)      379.50      (4.3%)   -2.6% ( -12% -    8%) 0.118
                 LowSloppyPhrase       31.28      (9.3%)       30.50      (8.8%)   -2.5% ( -18% -   17%) 0.380
        AndHighHighDayTaxoFacets       12.43      (5.0%)       12.20      (4.6%)   -1.8% ( -10% -    8%) 0.225
       BrowseDayOfYearSSDVFacets        2.53     (14.2%)        2.49     (12.0%)   -1.7% ( -24% -   28%) 0.681
            MedTermDayTaxoFacets       16.95      (5.8%)       16.67      (7.4%)   -1.6% ( -13% -   12%) 0.436
                          Fuzzy2       40.90      (8.5%)       40.26     (10.0%)   -1.6% ( -18% -   18%) 0.595
                     MedSpanNear      123.79     (12.1%)      122.08     (13.0%)   -1.4% ( -23% -   26%) 0.727
                   OrHighNotHigh      615.22      (7.2%)      607.15      (6.4%)   -1.3% ( -13% -   13%) 0.540
                      AndHighLow      842.03      (3.1%)      831.74      (3.2%)   -1.2% (  -7% -    5%) 0.220
                         Prefix3      877.74      (2.9%)      867.66      (3.4%)   -1.1% (  -7% -    5%) 0.246
                      HighPhrase       59.19      (4.5%)       58.62      (4.1%)   -1.0% (  -9% -    8%) 0.476
                     LowSpanNear      153.60      (6.1%)      152.29      (6.6%)   -0.9% ( -12% -   12%) 0.671
                       OrHighLow      670.48      (5.0%)      664.93      (4.5%)   -0.8% (  -9% -    9%) 0.583
               HighTermMonthSort     1357.59      (3.2%)     1350.69      (5.6%)   -0.5% (  -9% -    8%) 0.726
           BrowseMonthTaxoFacets        2.02      (7.5%)        2.02      (8.4%)   -0.1% ( -14% -   17%) 0.964
            HighIntervalsOrdered       20.84      (8.3%)       20.83      (8.6%)   -0.1% ( -15% -   18%) 0.978
                          IntNRQ      147.02      (1.6%)      146.94      (1.3%)   -0.1% (  -2% -    2%) 0.912
                        Wildcard      173.40      (2.9%)      173.51      (2.6%)    0.1% (  -5% -    5%) 0.942
     BrowseRandomLabelTaxoFacets        1.40      (3.5%)        1.40      (3.3%)    0.3% (  -6% -    7%) 0.803
                      AndHighMed      310.68      (8.0%)      311.55      (8.2%)    0.3% ( -14% -   17%) 0.912
         AndHighMedDayTaxoFacets       24.35      (4.0%)       24.44      (3.2%)    0.4% (  -6% -    7%) 0.737
                 MedSloppyPhrase       26.19      (2.7%)       26.31      (3.7%)    0.4% (  -5% -    7%) 0.668
                    HighSpanNear        8.56      (4.7%)        8.61      (4.9%)    0.5% (  -8% -   10%) 0.737
                     AndHighHigh      141.75     (19.2%)      142.51     (17.5%)    0.5% ( -30% -   46%) 0.927
                       LowPhrase      204.22      (3.4%)      205.45      (3.5%)    0.6% (  -6% -    7%) 0.583
                    OrNotHighMed      591.68      (5.7%)      595.28      (6.7%)    0.6% ( -11% -   13%) 0.758
                      TermDTSort      219.14      (3.9%)      220.82      (5.6%)    0.8% (  -8% -   10%) 0.615
                HighSloppyPhrase       21.66      (4.6%)       21.85      (7.2%)    0.9% ( -10% -   13%) 0.638
                    OrNotHighLow      838.61      (4.7%)      847.01      (4.8%)    1.0% (  -8% -   10%) 0.503
                         Respell       20.48      (6.6%)       20.69      (5.5%)    1.1% ( -10% -   14%) 0.583
                          Fuzzy1       45.98      (8.0%)       46.50      (8.0%)    1.1% ( -13% -   18%) 0.653
                      OrHighHigh      109.85     (20.7%)      111.12     (20.8%)    1.2% ( -33% -   53%) 0.860
     BrowseRandomLabelSSDVFacets        1.59      (7.1%)        1.61     (10.4%)    1.3% ( -15% -   20%) 0.653
                       MedPhrase      421.59      (3.2%)      427.00      (3.2%)    1.3% (  -4% -    7%) 0.204
                       OrHighMed      316.82     (13.9%)      321.65     (13.6%)    1.5% ( -22% -   33%) 0.726
                        HighTerm      411.86     (21.1%)      418.48     (20.1%)    1.6% ( -32% -   54%) 0.805
            HighTermTitleBDVSort       13.60      (5.3%)       13.84      (5.7%)    1.8% (  -8% -   13%) 0.312
            BrowseDateTaxoFacets        1.82      (6.3%)        1.85      (7.3%)    1.8% ( -11% -   16%) 0.408
               HighTermTitleSort      102.22      (4.6%)      104.06      (4.9%)    1.8% (  -7% -   11%) 0.232
       BrowseDayOfYearTaxoFacets        1.87      (6.8%)        1.90      (8.3%)    1.8% ( -12% -   18%) 0.446
                   OrNotHighHigh      544.66     (10.8%)      554.63     (12.4%)    1.8% ( -19% -   28%) 0.619
          OrHighMedDayTaxoFacets        5.43      (7.4%)        5.53      (7.2%)    1.9% ( -11% -   17%) 0.417
                           range     3046.35      (8.8%)     3106.25      (8.5%)    2.0% ( -14% -   21%) 0.472
                         LowTerm      866.65     (31.5%)      884.09     (30.8%)    2.0% ( -45% -   94%) 0.838
                        PKLookup      133.59      (8.8%)      137.11      (8.0%)    2.6% ( -13% -   21%) 0.321
                    OrHighNotLow      459.18     (11.2%)      472.18      (9.2%)    2.8% ( -15% -   26%) 0.382
             MedIntervalsOrdered       77.62      (9.3%)       80.07      (8.9%)    3.2% ( -13% -   23%) 0.274
            BrowseDateSSDVFacets        0.54     (11.9%)        0.56     (10.9%)    3.5% ( -17% -   29%) 0.337
             LowIntervalsOrdered       20.99      (6.9%)       21.72      (7.2%)    3.5% (  -9% -   18%) 0.117
                         MedTerm      394.43     (11.2%)      409.52     (14.7%)    3.8% ( -19% -   33%) 0.356
                          IntSet      373.81      (8.4%)      388.84     (10.3%)    4.0% ( -13% -   24%) 0.177
           BrowseMonthSSDVFacets        2.68     (13.7%)        2.84     (19.3%)    5.8% ( -23% -   45%) 0.272
```

msokolov commented:
thanks for checking w/smaller K/N - glad there was no regression there
